src/array/cpu/id_hash_map.h

```diff
@@ -0,0 +1,141 @@
+/**
+ *  Copyright (c) 2023 by Contributors
+ * @file array/cpu/id_hash_map.h
+ * @brief Class about id hash map
+ */
+
+#ifndef DGL_ARRAY_CPU_ID_HASH_MAP_H_
+#define DGL_ARRAY_CPU_ID_HASH_MAP_H_
+
+#include <dgl/aten/types.h>
+#include <dgl/runtime/device_api.h>
+
+#include <memory>
+#include <vector>
+
+namespace dgl {
+namespace aten {
+
+/**
+ * @brief A CPU targeted hashmap for mapping duplicate and non-consecutive ids
+ * in the provided array to unique and consecutive ones. It utilizes
+ * multi-threading to accelerate the insert and search speed. Currently it is
+ * only designed to be used in `ToBlockCpu` for optimizing.
+ *
+ * The hashmap should be used in two phases. With the first being creating the
+ * hashmap, and then init it with an id array. After that, searching any old ids
+ * to get the mappings according to your need.
+ *
+ * For example, for an array A with following entries:
+ * [98, 98, 100, 99, 97, 99, 101, 100, 102]
+ * Create the hashmap H with:
+ * `H = CpuIdHashMap()` (1)
+ * And Init it with:
+ * `U = H.Init(A)` (2)  (U is an id array used to store the unqiue
+ * ids in A).
+ * Then U should be (U is not exclusive as the element order is not
+ * guaranteed to be steady):
+ * [98, 100, 99, 97, 101, 102]
+ * And the hashmap should generate following mappings:
+ *  * [
+ *   {key: 98, value: 0},
+ *   {key: 100, value: 1},
+ *   {key: 99, value: 2},
+ *   {key: 97, value: 3},
+ *   {key: 101, value: 4},
+ *   {key: 102, value: 5}
+ * ]
+ * Search the hashmap with array I=[98, 99, 102]:
+ * R = H.Map(I) (3)
+ * R should be:
+ * [0, 2, 5]
+ **/
+template <typename IdType>
+class IdHashMap {
+ public:
+  /**
+   * @brief An entry in the hashtable.
+   */
+  struct Mapping {
+    /**
+     * @brief The ID of the item inserted.
+     */
+    IdType key;
+    /**
+     * @brief The value of the item inserted.
+     */
+    IdType value;
+  };
+
+  /**
+   * @brief Cross platform CAS operation.
+   * It is an atomic operation that compares the contents of a memory
+   * location with a given value and, only if they are the same, modifies
+   * the contents of that memory location to a new given value.
+   *
+   * @param ptr The pointer to the object to test and modify .
+   * @param old_val The value expected to be found in `ptr`.
+   * @param new_val The value to store in `ptr` if it is as expected.
+   *
+   * @return Old value pointed by the `ptr`.
+   */
+  static IdType CompareAndSwap(IdType* ptr, IdType old_val, IdType new_val);
```

Comment:
I don't think this is the right place for this function since CAS can be applicable to other components as well.  Maybe move it to an inline function in `dgl::runtime` namespace.

Reply:
It is a public static method so it can be reused in other modulars.